### PR TITLE
Resolve ImmutableOpenMap issue from core refactor

### PR DESCRIPTION
**Pull request by scrawfor99**
### Description
Removes use of the ImmutableOpenMap from Security code to resolve issue #2714. 

### Check List
- [ ] ~New functionality includes testing~
- [ ] ~New functionality has been documented~
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by reta — DISMISSED** [review]

**Comment by scrawfor99** [comment]
Will have to wait until the next build: https://build.ci.opensearch.org/job/distribution-build-opensearch/7748/ before we can actually confirm this is fixed. Most recent build gets pulled as part of build but that does not include the Core changes.

**Review by RyanL1997 — DISMISSED** [review]

File: src/integrationTest/java/org/opensearch/test/framework/matcher/AliasExistsMatcher.java
Comment (RyanL1997) [inline src/integrationTest/java/org/opensearch/test/framework/matcher/AliasExistsMatcher.java:44]: I think for this `getAliases()`, we also need to fix it, because it is also depending on the stale library of `ImmutableOpenMap`

**Review by RyanL1997 — COMMENTED** [review]
Hi @scrawfor99 , thanks for this fix. Sorry I had to dismissed my previous approve, because I was trying this fix on my local machine, and I think we need to fix some of the functions. I left a comment down below.

File: src/integrationTest/java/org/opensearch/test/framework/matcher/AliasExistsMatcher.java
Comment (scrawfor99) [reply src/integrationTest/java/org/opensearch/test/framework/matcher/AliasExistsMatcher.java:44]: So I took a look into this and I think it will actually be resolved once the new build is done. For whatever reason, the build with the core change is still not complete: https://build.ci.opensearch.org/job/distribution-build-opensearch/7750/

**Review by peternied — APPROVED** [review]

File: src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java
Comment (peternied) [inline src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java:652]:
```
2023-04-24T19:23:06.3310228Z   1> java.lang.ClassCastException: class java.util.Collections$UnmodifiableCollection cannot be cast to class java.util.Iterator (java.util.Collections$UnmodifiableCollection and java.util.Iterator are in module java.base of loader 'bootstrap')
2023-04-24T19:23:06.3310965Z   1> 	at org.opensearch.security.privileges.PrivilegesEvaluator$1.iterator(PrivilegesEvaluator.java:652) ~[main/:?]
```

Think you might be missing a `.iterator()` call before attempting to return, can you do this without the unchecked suppression?

You can repro the test failure with `org.opensearch.security.IndexIntegrationTests.testIndexResolveMinus` for a quick turnaround

**Review by RyanL1997 — APPROVED** [review]

File: src/integrationTest/java/org/opensearch/test/framework/matcher/GetSettingsResponseContainsIndicesMatcher.java
Comment (cwperks) [inline src/integrationTest/java/org/opensearch/test/framework/matcher/GetSettingsResponseContainsIndicesMatcher.java:41]: Can this be simplified to:

```suggestion
		final Set<String> indexKeys = new HashSet<>();
		for (ObjectObjectCursor<String, Settings> cursor : response.getIndexToSettings()) {
			indexKeys.add(cursor.key);
		}
```

Comment (cwperks) [reply src/integrationTest/java/org/opensearch/test/framework/matcher/GetSettingsResponseContainsIndicesMatcher.java:41]: Looks like either would work, but since we are not using the values in the map and only getting a keySet it would take less memory.

**Review by cwperks — COMMENTED** [review]
Thank you @scrawfor99, the changes look good to me. I left one suggestion so let me know what you think, it wouldn't block this PR from being approved and merged. If the suggestion doesn't work then let me know and I will approve this PR.

**Review by peternied — APPROVED** [review]

File: src/integrationTest/java/org/opensearch/test/framework/matcher/GetSettingsResponseContainsIndicesMatcher.java
Comment (scrawfor99) [reply src/integrationTest/java/org/opensearch/test/framework/matcher/GetSettingsResponseContainsIndicesMatcher.java:41]: Sorry Craig, I didn't see this before the PR got merged. I had left it as a map because that is what the original implementation was. I agree that a Set should work fine here as well. If you wanted to swap it you could open up a quick PR to change it in main.

File: src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java
Comment (andrross) [inline src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java:677]: Why did you make this change? Doesn't `indexMetaData.getAliases()` still return an ImmutableOpenMap? Seems like you should avoid making a copy of the map in any case.

Comment (scrawfor99) [reply src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java:677]: Hi @andrross, I made this swap because we could not use `keysIt` anymore. We needed `aliases` to be a normal map to then grab and iterator over the keyset.

Comment (andrross) [reply src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java:677]: @nknize What's happening here? Pretty sure making a copy of the the map is the exact opposite of your intention with this change.

Comment (scrawfor99) [reply src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java:677]: This was the issue that arose when the change was made to core: https://github.com/opensearch-project/security/issues/2714. I just made changes to get things Green. Sorry if this was not the intended implementation...

Comment (andrross) [reply src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java:677]: Sorry @scrawfor99! Definitely not trying to blame you here! I don't understand why `keysIt` doesn't work anymore, and I don't think the intent with @nknize's change is to force downstream dependencies to copy data structures when doing a simple iteration.

Comment (nknize) [reply src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java:677]: No copies unless you need to make changes to the map (even then the core should return an `unmodifiableMap` so you would only be able to change a clone). 

If you just need an iterator over the keys then use `.keySet().iterator()`

**Pull request merged**
